internal/operator-controller/rukpak/bundle/config.go

```diff
@@ -6,119 +6,498 @@ import (
+	} else {
+		// allow null or valid namespace string
+		stringSchema := map[string]interface{}{
+			"type":   "string",
+			"format": format,
 		}
+		watchNSProperty["anyOf"] = []map[string]interface{}{
+			{"type": "null"},
+			stringSchema,
+		}
+	}
+
+	return watchNSProperty
+}
+
+// selectNamespaceFormat chooses the appropriate format validator based on install modes.
+//
+//   - When OwnNamespace only -> "installNamespace" (must equal install namespace)
+//   - When SingleNamespace only -> "notInstallNamespace" (must NOT equal install namespace)
+//   - Otherwise -> "k8sNamespace" (any valid namespace)
+func selectNamespaceFormat(hasOwnNamespace, hasSingleNamespace bool) string {
+	if hasOwnNamespace && !hasSingleNamespace {
+		return "installNamespace"
+	}
+	if hasSingleNamespace && !hasOwnNamespace {
+		return "notInstallNamespace"
+	}
+	return "k8sNamespace"
+}
+
+// validateConfigWithSchema validates the config bytes against the generated JSON schema.
+//
+// Note: We create a new compiler on each call because the custom formats (installNamespace,
+// notInstallNamespace) are closures that capture the installNamespace value, which varies
+// per ClusterExtension. If this becomes a performance bottleneck, we could cache compiled
+// schemas keyed by (schema JSON + installNamespace), but for now the overhead is negligible.
+func validateConfigWithSchema(configBytes []byte, schemaJSON string, installNamespace string) error {
+	var configData interface{}
+	if err := yaml.Unmarshal(configBytes, &configData); err != nil {
+		return formatUnmarshalError(err)
 	}
 
-	// only accept install namespace if OwnNamespace install mode is supported
-	if config.WatchNamespace != nil && *config.WatchNamespace == installNamespace &&
-		!bundleInstallModeSet.Has(v1alpha1.InstallMode{Type: v1alpha1.InstallModeTypeOwnNamespace, Supported: true}) {
-		return fmt.Errorf("invalid 'watchNamespace' %q: must not be install namespace (%s)", *config.WatchNamespace, installNamespace)
+	var schemaData interface{}
+	if err := json.Unmarshal([]byte(schemaJSON), &schemaData); err != nil {
+		return fmt.Errorf("failed to parse schema: %w", err)
 	}
 
-	// only accept non-install namespace is SingleNamespace is supported
-	if config.WatchNamespace != nil && *config.WatchNamespace != installNamespace &&
-		!bundleInstallModeSet.Has(v1alpha1.InstallMode{Type: v1alpha1.InstallModeTypeSingleNamespace, Supported: true}) {
-		return fmt.Errorf("invalid 'watchNamespace' %q: must be install namespace (%s)", *config.WatchNamespace, installNamespace)
+	compiler := jsonschema.NewCompiler()
+
+	// Register custom formats for validation
+	compiler.RegisterFormat(&jsonschema.Format{
+		Name:     "k8sNamespace",
+		Validate: validateK8sNamespace,
+	})
+	compiler.RegisterFormat(&jsonschema.Format{
+		Name: "installNamespace",
+		Validate: func(value interface{}) error {
+			// First validate it's a valid namespace name
+			if err := validateK8sNamespace(value); err != nil {
```

Comment:
I wonder if this is necessary here since the install namespace is provided by the system and expected to be a valid namespace name. It could also mean the user goes through two rounds of invalid configuration: first the namespace name isn't valid, then finds out it cannot be the install namespace. I don't think it's the end of the world, just a thought...wdyt?

Reply:
Marking as resolved to facilitate the review